packages/sprintf/types/index.d.ts

```diff
@@ -1,87 +1,267 @@
-			: ExtractPositionalPlaceholders<Rest>
-		:  unknown[];
-
-type ExtractStaticPrecisionPlaceholders<T extends string> =
-	StripEscapedPercents<T> extends `${any}%.${infer Precision extends number}${infer Spec}${infer Rest}`
-		? Spec extends S
-			? [Specifiers[Spec], ...ExtractStaticPrecisionPlaceholders<Rest>]
-			: never
-		: [];
+// ---- Positional Placeholder Parsing ----
 
-type ExtractDynamicPrecisionPlaceholder<T extends string> =
-	StripEscapedPercents<T> extends `${any}%.*${infer Spec}${infer Rest}`
-		? Spec extends S
-			? [number, Specifiers[Spec], ...ExtractDynamicPrecisionPlaceholder<Rest>]
+// Extracts a numeric position from format like "2$s"
+type ExtractPositionalNumber<T extends string> =
+	T extends `${infer NumStr}$${infer Rest}`
+		? NumStr extends `${number}`
+			? [NumStr, Rest]
 			: never
+		: never;
+
+// Recursively collects all positional placeholders and their expected types
+type CollectPositionalPlaceholders<
+	T extends string,
+	Collected extends Record<string, any> = {},
+> =
+	StripEscapedPercents<T> extends `${infer _}%${infer AfterPercent}`
+		? ExtractPositionalNumber<AfterPercent> extends never
+			? Collected
+			: ExtractPositionalNumber<AfterPercent> extends [
+						infer PosStr extends string,
+						infer AfterPos extends string,
+				  ]
+				? ParsePrecisionAndSpec<AfterPos> extends [
+						infer Precision extends string,
+						infer Spec extends S,
+						infer Rest extends string,
+					]
+					? IsValidSpec<Spec> extends never
+						? never
+						: Precision extends '.*'
+							? CollectPositionalPlaceholders<
+									Rest,
+									Collected & { [K in PosStr]: [number, Specifiers[Spec]] }
+								>
+							: CollectPositionalPlaceholders<
+									Rest,
+									Collected & { [K in PosStr]: Specifiers[Spec] }
+								>
+					: CollectPositionalPlaceholders<AfterPos, Collected>
+				: Collected
+		: Collected;
+
+// ---- Positional Argument Extraction ----
+
+// Gets the max index used in positional placeholders
+type GetMaxPosition<T extends Record<string, any>> = {
+	[K in keyof T]: K extends `${infer N extends number}` ? N : never;
+}[keyof T];
+
+// Computes max of a number by building up a tuple
+type Max<N extends number, A extends any[] = []> = [N] extends [
+	Partial<A>['length'],
+]
+	? A['length']
+	: Max<N, [0, ...A]>;
+
+// Builds a final tuple of arguments from the collected positional types
+type BuildPositionalTuple<
+	Collected extends { [K: number]: any },
+	MaxPos extends number,
+	CurrentPos extends number = 1,
+	Result extends any[] = [],
+> =
+	CurrentPos extends Add<MaxPos, 1>
+		? Result
+		: `${CurrentPos}` extends keyof Collected
+			? Collected[`${CurrentPos}`] extends [any, any]
+				? BuildPositionalTuple<
+						Collected,
+						MaxPos,
+						Add<CurrentPos, 1>,
+						[...Result, ...Collected[CurrentPos]]
+					>
+				: BuildPositionalTuple<
+						Collected,
+						MaxPos,
+						Add<CurrentPos, 1>,
+						[...Result, Collected[CurrentPos]]
```

Comment:
Should these also be converting `CurrentPos` to string like you're doing in the checks above?


```suggestion
						[...Result, ...Collected[`${CurrentPos}`]]
					>
				: BuildPositionalTuple<
						Collected,
						MaxPos,
						Add<CurrentPos, 1>,
						[...Result, Collected[`${CurrentPos}`]]
```

Reply:
I think we can change the first one to [CurrentPos] as Collected is [key: number]: any
somehow I missed that.